src/cmd/ksh93/tests/modifiers.sh

```diff
@@ -0,0 +1,42 @@
+# These are the tests for various style of parameter expansion modifiers supported by ksh93
+
+# Check ${parameter:-word} style of parameter expansion. ':' is optional and is used to check for null (empty string).
+unset foo
+[[ ${foo:-bar} == bar ]]  || log_error  '${foo:-bar} not bar when foo is not set'
+
+unset foo
+[[ ${foo-bar} == bar ]]  || log_error  '${foo-bar} not bar when foo is not set'
+
+foo=""
+[[ ${foo-bar} == "" ]]  || log_error  '${foo-bar} not "" when foo is null'
+
+# Check ${parameter:=word} style of parameter expansion. ':' is optional and is used to check for null (empty string).
+unset foo
+[[ ${foo:=bar} == bar ]]  || log_error '${foo:=bar} not bar when foo is not set'
+
+unset foo
+[[ ${foo=bar} == bar ]]  || log_error '${foo=bar} not bar when foo is not set'
+
+foo=""
+[[ ${foo=bar} == "" ]]  || log_error  '${foo=bar} not "" when foo is null'
+
+# Check ${parameter:?word} style of parameter expansion. ':' is optional and is used to check for null (empty string).
+[[ $( $SHELL -c 'unset foo; print ${foo:?bar not set}' 2>&1 ) =~ "bar not set" ]]  || log_error '${foo:?bar} does not display error if foo is unset'
```

Comment:
FYI, for complex situations like this I prefer this pattern so that the test log provides more useful data:

```
expect="bar not set"
actual=$( $SHELL -c 'unset foo; print ${foo:?bar not set}' 2>&1 )
if [[ $actual =~ $expect ]] || log_error 'message' "$expect" "$actual"
```